syntax = "proto3";

package stats;

message PostStatsResponse {
    int64 views = 1;
    int64 likes = 2;
    int64 comments = 3;
}

message PostStatsRequest {
    string post_id = 1;
}

message DayStats {
    string date = 1;
    int64 stat = 2;
}

message PostHistoryResponse {
    repeated DayStats history = 1;
}

enum SortParam {
    VIEWS = 0;
    LIKES = 1;
    COMMENTS = 2;
}

message TopTenPostsRequest {
    SortParam param = 1;
}

message TopTenUsersRequest {
    SortParam param = 1;
}

message TopTenPostsResponse {
    repeated string post_ids = 1;
}

message TopTenUsersResponse {
    repeated string user_ids = 1;
}

service StatsService {
    rpc GetPostStats (PostStatsRequest) returns (PostStatsResponse);
    rpc GetPostViewsHistory (PostStatsRequest) returns (PostHistoryResponse);
    rpc GetPostLikesHistory (PostStatsRequest) returns (PostHistoryResponse);
    rpc GetPostCommentsHistory (PostStatsRequest) returns (PostHistoryResponse);
    rpc GetPostRecentComments (PostStatsRequest) returns (PostHistoryResponse);
    rpc GetTopTenPosts (TopTenPostsRequest) returns (TopTenPostsResponse);
    rpc GetTopTenUsers (TopTenUsersRequest) returns (TopTenUsersResponse);
}